src/Transform/ONNX/ONNXHybridTransformPass.cpp

```diff
@@ -0,0 +1,147 @@
+  using OpInterfaceRewritePattern<ShapeInference>::OpInterfaceRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      ShapeInference shapeOp, PatternRewriter &rewriter) const override {
+    // TODO: check if it's necessary to skip ops that satisfy
+    // !returnsDynamicOrUnknownShape (see ShapeInferencePass.cpp)
+
+    // Verify the operation before attempting to infer the shape of the
+    // produced output(s).
+    Optional<RegisteredOperationName> registeredInfo =
+        shapeOp->getName().getRegisteredInfo();
+    if (registeredInfo && failed(registeredInfo->verifyInvariants(&*shapeOp)))
+      return shapeOp.emitOpError("verification failed");
+
+    // Infer the results shapes.
+    if (failed(shapeOp.inferShapes([](Region &region) {})))
+      return shapeOp.emitOpError("shape inference failed");
+
+    return success();
+  }
+};
+
+struct ReturnShapesPattern : public OpRewritePattern<ONNXReturnOp> {
+  using OpRewritePattern<ONNXReturnOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      ONNXReturnOp returnOp, PatternRewriter &rewriter) const override {
+    Operation *parent = returnOp->getParentOp();
+    (void)parent;
+    assert(parent && "every onnx.Return op has a parent");
+    if (auto shapeOp = llvm::dyn_cast<ShapeInference>(parent)) {
+      if (failed(shapeOp.inferShapes([](Region &region) {})))
+        return shapeOp.emitOpError("shape inference failed");
+    } else {
+      llvm_unreachable("onnx.Return always has if/loop/scan parent");
+    }
+    return success();
+  }
+};
+
+// The pass combines patterns for shape inference and other ONNX-to-ONNX
+// transforms, controlled by the shapeInferenceOnly constructor argument.
+//
+// Shape inference is done with two patterns. One pattern is for regular ONNX
+// ops from the ONNX specification, which all implement the ShapeInference
+// interface. The other pattern is for ONNXReturnOp which terminates
+// if/loop/scan subgraphs. The pass executes patterns top down so shape
+// inference cascades through the ops from the graph's inputs to outputs, and
+// recursively into subgraphs. Ops with subgraphs, namely if/loop/scan, are
+// matched by the first pattern before the pass recurses into the subgraph. The
+// recursive subgraph pass ends with the ONNXReturnOp whose pattern reruns shape
+// inference for the parent if/loop/scan op. The effect is that the two runs of
+// the parent if/loop/scan op each accomplishes one half of shape propagation to
+// and from the subgraph: The first run propagates input shapes from the parent
+// op to the subgraph and the second run propagates result shapes from the
+// subgraph to the parent op.
+struct ONNXHybridTransformPass
+    : public PassWrapper<ONNXHybridTransformPass, OperationPass<func::FuncOp>> {
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(ONNXHybridTransformPass)
+
+  ONNXHybridTransformPass(bool shapeInferenceOnly)
+      : shapeInferenceOnly(shapeInferenceOnly) {}
```

Comment:
I am think of a  general control on which transformation pattern, not just shape inference only, will be added in the hybrid pass.